src/GradhSph/GradhSphTree.cpp

```diff
@@ -427,46 +393,19 @@ void GradhSphTree<ndim,ParticleType>::UpdateAllSphHydroForces
 
         bool do_hydro = sph->types[activepart[j].ptype].hydro_forces ;
         if (do_hydro){
-          Typemask hydromask  = sph->types[activepart[j].ptype].hydromask;
 
-          for (k=0; k<ndim; k++) rp[k] = activepart[j].r[k];
-          hrangesqdi = activepart[j].hrangesqd;
-          Nhydroaux = 0;
+          // Ask to the neighbour manager for the list of neighbours
+          int* sphlist;
+          ParticleType<ndim>* neibpart;
+          FLOAT* drmag;
+          FLOAT* dr;
 
-          // Validate that gather neighbour list is correct
-#if defined(VERIFY_ALL)
-          if (neibcheck) this->CheckValidNeighbourList(i, sph->Nhydro + sph->NPeriodicGhost,
-                                                       Nneib, neiblist, sphdata, "all");
-#endif
-
-          // Compute distances and the inverse between the current particle and all neighbours here,
-          // for both gather and inactive scatter neibs.  Only consider particles with j > i to
-          // compute pair forces once unless particle j is inactive.
-          //-----------------------------------------------------------------------------------------
-          for (jj=0; jj<Nneib; jj++) {
-
-            // Skip non-hydro particles and the current active particle.
-            if (hydromask[neibpart[jj].ptype] == false) continue;
-
-            for (k=0; k<ndim; k++) draux[k] = neibpart[jj].r[k] - rp[k];
-            drsqd = DotProduct(draux, draux, ndim) + small_number;
-
-            if (drsqd <= small_number) continue ;
+          Typemask hydromask  = sph->types[activepart[j].ptype].hydromask;
 
-            // Compute relative position and distance quantities for pair
-            if (drsqd <= hrangesqdi || drsqd <= neibpart[jj].hrangesqd) {
-              drmag[Nhydroaux] = sqrt(drsqd);
-              invdrmag[Nhydroaux] = (FLOAT) 1.0/drmag[Nhydroaux];
-              for (k=0; k<ndim; k++) dr[Nhydroaux*ndim + k] = draux[k]*invdrmag[Nhydroaux];
-              levelneib[neiblist[jj]] = max(levelneib[neiblist[jj]],activepart[j].level);
-              sphlist[Nhydroaux] = jj;
-              Nhydroaux++;
-            }
-          }
-          //-----------------------------------------------------------------------------------------
+          const int Nneib=neibmanager.GetNeibList(activepart[j],hydromask,&sphlist,&neibpart,&drmag,&dr,levelneib);
 
           // Compute all neighbour contributions to hydro forces
-          sph->ComputeSphHydroForces(i,Nhydroaux,sphlist,drmag,invdrmag,dr,activepart[j],neibpart);
+          sph->ComputeSphHydroForces(i,Nneib,sphlist,drmag,NULL,dr,activepart[j],neibpart);
```

Comment:
Do we really need to pass in the vectors drmag and dr to ComputeSphHydroForces? I can see why we don't want to recompute them. But the meshless for example as far as I can see doesn't use them (although, they are passed in! I guess this comes from copying and pasting).
Maybe the best thing to do is to make neibpart of a different type than the whole particle. And dr and drmag can be saved inside there, so that we don't need these supporting arrays which are annoying.